plugins/newsdownloader.koplugin/internaldownloadbackend.lua

```diff
@@ -0,0 +1,49 @@
+local http = require("socket.http")
+local https = require("ssl.https")
+local logger = require("logger")
+local ltn12 = require("ltn12")
+local socket = require('socket')
+local socket_url = require("socket.url")
+
+local InternalDownloadBackend = {}
+local max_redirects = 10;
+
+function InternalDownloadBackend:getResponseAsString(url, redirectCount)
+    if not redirectCount then
+        redirectCount = 0
+    elseif redirectCount == max_redirects then
+        error("InternalDownloadBackend: reached max redirects: ", redirectCount)
+    end
+    logger.dbg("InternalDownloadBackend: url :", url)
+    local request, sink = {}, {}
+    request['sink'] = ltn12.sink.table(sink)
+    request['url'] = url
+    local parsed = socket_url.parse(url)
+
+    local httpRequest = parsed.scheme == 'http' and http.request or https.request
+    -- first argument returned by skip is code
+    local _, headers, status = socket.skip(1, httpRequest(request))
+
+    if status ~= "HTTP/1.1 200 OK" then
+        logger.dbg("InternalDownloadBackend: HTTP response code <> 200. Response code: ", status)
+        if status and string.sub(status, 1, 1) ~= "3" then -- handle 301, 302...
+           if headers and headers["location"] then
+              local redirected_url = headers["location"]
+              logger.dbg("InternalDownloadBackend: Redirecting to url: ", redirected_url)
+              return self:getResponseAsString(redirected_url, redirectCount + 1)
+           end
+        else
+           error("InternalDownloadBackend: Don't know how to handle HTTP status:", status)
+        end
+    end
+    return table.concat(sink)
+end
+
+function InternalDownloadBackend:download(url, path)
+   local response = self:getResponseAsString(url)
+   local file = assert(io.open(path, 'w'))
```

Comment:
Note that an assert means insta-crash in case it fails. Which is excellent during development, but not so much for regular use. ;-)

There are a couple of methods we've added to get around this issue:

```lua
local dbg = require("dbg")
if dbg.is_on then
    -- do some asserting or erroring
end
```

Example use:

https://github.com/koreader/koreader/blob/43a6cf4b2eb04fb8826107a7a38b1fa5cc072caf/frontend/ui/size.lua#L75-L94

Then there's also `dbg.guard` and `dbg.dassert`.

```lua
dbg.dassert(file ~= nil)
```

https://github.com/koreader/koreader/blob/775b7a364f4b0a1b82fc236d4aaa6513922f9904/frontend/ui/uimanager.lua#L298-L302

Reply:
ok I'll remove assert